simpleast-core/src/main/java/com/discord/simpleast/code/TypeScript.kt

```diff
@@ -0,0 +1,249 @@
+         * set internals()
+         * ```
+         */
+         private val PATTERN_TYPESCRIPT_FUNC = 
+             """^((?:function\*?|static|get|set|async)\s)(\s*[a-zA-Z_$][a-zA-Z0-9_$]*)?(\s*<.*>)?""".toRegex(RegexOption.DOT_MATCHES_ALL).toPattern()
+
+         fun <RC, S> createFunctionRule(codeStyleProviders: CodeStyleProviders<RC>) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_TYPESCRIPT_FUNC) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S): ParseSpec<RC, S> {
+              val pre = matcher.group(1)
+              val signature = matcher.group(2)
+              val generics = matcher.group(3)
+              return ParseSpec.createTerminal(FunctionNode(pre!!, signature, generics, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  class FieldNode<RC>(
+      definition: String, name: String,
+      codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      StyleNode.TextStyledNode(definition, codeStyleProviders.keywordStyleProvider),
+      StyleNode.TextStyledNode(name, codeStyleProviders.identifierStyleProvider),
+  ) {
+    companion object {
+      /**
+       * Matches against a TypeScript field definition.
+       *
+       * ```
+       * var x = 1;
+       * let y = 5;
+       * const z = 10;
+       * const h: string = 'Hello world';
+       * ```
+       */
+      private val PATTERN_TYPESCRIPT_FIELD =
+          Pattern.compile("""^(var|let|const)(\s+[a-zA-Z_$][a-zA-Z0-9_$]*)""")
+
+      fun <RC, S> createFieldRule(
+          codeStyleProviders: CodeStyleProviders<RC>
+      ) =
+          object : Rule<RC, Node<RC>, S>(PATTERN_TYPESCRIPT_FIELD) {
+            override fun parse(matcher: Matcher, parser: Parser<RC, in Node<RC>, S>, state: S):
+                ParseSpec<RC, S> {
+              val definition = matcher.group(1)
+              val name = matcher.group(2)
+              return ParseSpec.createTerminal(
+                  FieldNode(definition!!, name!!, codeStyleProviders), state)
+            }
+          }
+    }
+  }
+
+  class ObjectPropertyNode<RC>(
+      prefix: String, accessModifier: String?, property: String, suffix: String,
+      codeStyleProviders: CodeStyleProviders<RC>
+  ) : Node.Parent<RC>(
+      StyleNode.TextStyledNode(prefix, codeStyleProviders.defaultStyleProvider),
+      accessModifier?.let { StyleNode.TextStyledNode(accessModifier, codeStyleProviders.keywordStyleProvider) },
+      StyleNode.TextStyledNode(property, codeStyleProviders.identifierStyleProvider),
+      StyleNode.TextStyledNode(suffix, codeStyleProviders.defaultStyleProvider),
+  ) {
+    companion object {
+      /**
+       * Matches against a TypeScript object property.
+       *
+       * ```
+       * { foo: 'bar' }
+       * ```
+       */
+      private val PATTERN_TYPESCRIPT_OBJECT_PROPERTY = 
+          Pattern.compile("""^([\{\[\(\,;])(\s*(?:public|private|protected))?(\s*[a-zA-Z0-9_$]+)((?:\s*\?)?\s*:)""")
```

Comment:
Index signatures and [mapping modifiers](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#mapping-modifiers) also need to be supported.